### ISSUE TITLE
Crash during concurrent garbage collection

### ISSUE THREAD

**Issue by jkachmar** [NixOS/nix#7370]
### Description

`nix` builds intermittently fail with the following error message:

```
error: cannot connect to socket at '/nix/var/nix/gc-socket/socket': No such file or directory
```

This only appears to occur on systems with scheduled garbage collection and/or concurrent calls to collect garbage.

**UPDATE (29 March 2023)**: In addition to the above failure, we have also been seeing the following errors occur in close proximity to scheduled garbage collection; it's not entirely clear _why_ Nix might GC a store path that's a dependency of an active realization, but that does appear to be what's happening:

```
FileNotFoundError: [Errno 2] No such file or directory: '/nix/store/<NIX HASH>-<TARGET>'
```

### Reproduction

Unable to consistently reproduce.

### Expected Behavior

Nix should be able to do one of the following:
- perform concurrent garbage collection without crashing
- make concurrent garbage collection optional

### Nix Version

```
❯ nix --version
nix (Nix) 2.10.3
```

#### Related Issues & Code Paths

AFAICT this seems to have been introduced after #5149; at the very least, this problem did not seem to appear on systems running Nix 2.4 & has since appeared with some frequency on systems running Nix 2.10.

The error appears to be coming from this code path: https://github.com/NixOS/nix/blob/0f64bf445a229c4f5ce19dc19a552f58701ca1ca/src/libstore/gc.cc#L118-L121

#### Potential Causes

This is very much outside of my area of expertise, but my guess is that the following occurs:

- process A starts GC, takes the lock
- process B starts GC, sees that there is a lock, and attempts to communicate with the GC process via its GC socket
- process A releases the lock **before** process B can attempt to probe for the socket
- process B fails

I don't know that this matters at all, but looking at the following code block I would have expected Nix to call the GC socket with `send`/`recv` rather than `writeFull`/`readFull`:
https://github.com/NixOS/nix/blob/0f64bf445a229c4f5ce19dc19a552f58701ca1ca/src/libstore/gc.cc#L140-L161

**Comment by jkachmar** [comment]
cross-linking with #6757, on the off-chance that this is resolved by the fix described in that issue.

**Comment by thufschmitt** [comment]
I've seen this happen every once in a while on the CI some time ago. It seemed to have disappeared, so I expected that the issue got magically fixed.

I'm not entirely sure, but should we maybe catch more errors when trying to connect? At least handle the case where the socket doesn't exist any more since I guess that's what's causing the `No such file or directory`.

https://github.com/NixOS/nix/blob/0f64bf445a229c4f5ce19dc19a552f58701ca1ca/src/libstore/gc.cc#L128-L137

**Comment by elaforge** [comment]
I recently came across another error, the `nix-store -r` client fails with `error: unexpected end-of-file`.  The nix-daemon log has a `src/libstore/build/substitution-goal.cc:190: void nix::PathSubstitutionGoal::referencesValid(): Assertion `worker.store.isValidPath(i)' failed.` which is around the same time so it's probably the real error.

**Comment by domenkozar** [comment]
```
running auto-GC to free 16132497408 bytes
finding garbage collector roots...
error:
       error: cannot connect to socket at '/nix/var/nix/gc-socket/socket': No such file or directory
```

Using Nix 2.14.1

**Comment by domenkozar** [comment]
@edolstra this might be another common cause for EOF errors.

**Comment by jkachmar** [comment]
Updated the issue description with another error that we've been encountering at `$WORK` which I believe is related to this:

```
FileNotFoundError: [Errno 2] No such file or directory: '/nix/store/<NIX HASH>-<TARGET>'
```

It's not entirely clear _why_ Nix might GC a store path that's a dependency of an active realization, but (as best I can tell) this seems to be what's happening.

**Comment by dhess** [comment]
We recently started seeing this on our `aarch64-darwin` builders, sometime in the last week or two. Builds are failing randomly and unreproducibly, which is pretty frustrating.

**Comment by roberth** [comment]
Similar error in test in CI: https://github.com/NixOS/nix/actions/runs/5022907118/jobs/9006902362?pr=8370
Path in the message is different during testing, but otherwise equivalent.
```
error: cannot connect to socket at '/private/tmp/nix-build-nix-2.16.0pre20230519_33d9d96.drv-0/nix-test/tests/gc-non-blocking/var/nix/gc-socket/socket'
```

**Comment by elaforge** [comment]
We've been having this problem continuously since upgrading past 2.4 (looks like it was added in 2.5).  We eventually implemented a special system to try to detect crashes due to GC and retry, which we have gradually added regexes to over time.  I recently upped the retry count from 3 to 12, and we've cherry-picked various commits that aimed to improve the situation, to no avail.  So over the last year or so it's been constant churn trying to work around the problems.

All of that is to say, at what point is it worth considering to revert the concurrent GC feature?  Or at least include a flag to opt out of it?  At this point, beyond continuing to make the retry framework fancier over time, my best idea is to reinvent the big GC lock by wrapping all nix commands to take a lock file, but clearly it would be better if nix itself could do that.  As it once did!

It's been a long time since the feature was added, and I don't know how invasive it was, and how much work to bring back the old slower but non-crashing way.  So before trying the wrap everything strategy, I'd want to evaluate how much work to patch it back into nix.  Imagining that I do that and it's not much work, would upstream be interested?  To be honest since we have the retry framework now I'm not strongly motivated, but if it keeps needing more tweaks maybe it's a sunk-cost trap.

I've also noticed that now I need to GC several times to actually get everything, because it tends to stop in the middle claiming things are used.  So it could be we also need a retry framework wrapping nix-collect-garbage.  I don't know for sure if this is related to concurrent GC, but it seems likely.

**Comment by fricklerhandwerk** [comment]
Picked at it with @edolstra today. @edolstra will add catch the `ENOENT` case and try to test it with an [artificial synchronisation point](https://github.com/NixOS/nix/blob/f5e620bf2be7ed552962ab5b6637771d5a4d64d3/src/libstore/gc.cc#L775-L777). Another possible mitigation would be creating the socket file very early on, such as in the local store initialisation.

It's still not clear how exactly the issue arises. The time between acquiring a GC lock and creating a socket should be in the nanoseconds, and the only thing we can imagine that it would *rarely* happen if one spins up fresh machines very, very often.

**Comment by edolstra** [comment]
It's probably more likely when auto-GC is enabled.

#8544 should fix the race when the GC runs for the first time.

The `FileNotFoundError` errors reported above sound like a separate issue. It also looks like a Python error so it's coming from somewhere else?

**Comment by elaforge** [comment]
To be clear, the `cannot connect to socket at` error seems to be the rarest symptom.  "File not found" is much more comment.  While the example is python, it's not python specific, any builder can suffer from it.  Even nix internally will suddenly fail to find a `.drv`.

Of course these may all be separate issues, they're unified only by the fact that they happen when GC is going on.  If it would be helpful, I can troll through old logs to get a more specific list of errors and their frequencies.  In addition, I can probably reproduce these pretty easily.  I haven't specifically tried, but I see a lot of them every time GC happens when building is happening, so if I intentionally created that situation I'd probably see them too.  But it's only worth doing it if I have a plan to gather more info from it, so let me know if I can help.

Here are the symptoms from the top of my head, in frequency order:

- file not found during build or eval.  Happens for both out paths and drv paths.
- nix-collect-garbage crashes with "path in use"
- cannot connect to socket
- unexpected end-of-file (nix-daemon: Assertion worker.store.isValidPath(i)' failed)

**Comment by edolstra** [comment]
To diagnose the "file not found" errors, it's important to know exactly what the client is doing, in particular whether it's registering the paths it's building as temporary or permanent GC roots. For instance, if you do something like
```
nix-store -r $(nix-instantiate ...)
```
then the .drvs created by `nix-instantiate` are not GC roots and can be deleted.

**Comment by elaforge** [comment]
This happens during remote build, so perhaps the sequence is:

- client connects, sends wopQueryMissing
- client uploads missing paths, inputSrcs
- client sends wopBuildDerivation
- build starts, but now GC deletes some of the paths that were uploaded or deemed already present

When a remote build is in progress, since the drv being built is not actually present in the store, how does it get recognized as a GC root?

**Comment by fricklerhandwerk** [comment]
This sounds like a completely different issue than originally describe, no?

**Comment by elaforge** [comment]
Well, it's a crash during concurrent GC.  We don't know if it's all the same bug with multiple symptoms, or if concurrent gc has numerous bugs, one for each kind of error.

The socket error I consider low priority because it's so rare, but it looks like other people get it too.  Missing files are very common, but not clear if others are reporting them too.

I'll look through our logs to try to get some stats about error frequency.

Shall I make a separate issue "missing paths during concurrent garbage collection"?

**Comment by NobbZ** [comment]
> The socket error I consider low priority because it's so rare, but it looks like other people get it too. Missing files are very common, but not clear if others are reporting them too.

In #8550 I discovered that in the nix-daemons CGroup old `nix-daemon $number` instances survive even restarts of the service and with system uptime accumulate. The higher the number of those processes the more likely the socket error seems to po-up.

After force `kill -9`ing the "zombies" the socket problem doesn't pop up anymore.

This observation is currently based on a small sample size, as I have not yet found a way to artificially increase the number of processes in the CGroup.
